### PR TITLE
Snapshot import export component options

### LINKED ISSUE #4605: Add control over contents of exported and imported Snapshots
### Description

We offer the feature to "Export" Snapshots, this contains, in a single file, the `flows`, `credentials` and `environment variables` for a given Instance. Users can then _import_ that into another Application, or instance of FlowFuse.

The challenge here is that we have customers running multiple instances of FlowFuse, where one instance is entirely development-only, and the other is production-only, and they want to be able to transfer instances between. The difficulty here is that our DevOps Pipelines are limited to instances _within_ a single FlowFuse instance, and as such, users are having to resort to manual export/import.

The manual route works, but contains env vars from development environments, and overwrites production credentials when imported. As such, the following is required of our import/export functionality:

```[tasklist]
### Tasks
- [ ] Offer users the option to toggle on/off the inclusion of "Flows", "Credentials" and "Environment Variables" when exporting a Snapshot
- [ ] Offer users the option to toggle on/off the inclusion of "Flows", "Credentials" and "Environment Variables" when importing a Snapshot
- [ ] When including Environment Variables, we should offer the following: "Include Keys & Values", "Include Keys Only"
```

We already have a similar UX for the control of the contents in the "Duplicate Instance" here:

image

### Which customers would this be available to

Everyone - CE/Starter/Team/Enterprise

### Have you provided an initial effort estimate for this issue?

I have provided an initial effort estimate

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #4605

## Description

* Deletes duplicate (unused) vue components
  * frontend/src/pages/application/components/ExportProjectComponents.vue
  * frontend/src/pages/application/components/ImportProjectComponents.vue
  * Removes eslint exceptions
* Updates snapshot controller to support component options
* Update snapshot import/export API to support component options
* Updates frontend API to support component options
* Updates UI for snapshot import/export component options
  * Repurposes ExportInstanceComponents as ExportImportComponents
  * adds adds instance of repurposed ExportImportComponents to:
    * frontend/src/components/dialogs/SnapshotImportDialog.vue
    * frontend/src/pages/application/Snapshots/components/dialogs/SnapshotExportDialog.vue
* add unit tests for import/export snapshot component options

**Potential clean up**: move `SnapshotExportDialog.vue` to be more reusable, sitting along side `frontend/src/components/dialogs/SnapshotImportDialog.vue`

### Unit Tests added
```
  Snapshot controller
    exportSnapshot
      ✔ should export a full snapshot by default
      ✔ should export a snapshot without creds
      ✔ should export a snapshot without flows
      ✔ should export a snapshot without env (envVars: false)
      ✔ should export a snapshot without env values (envVars: "keys")
    uploadSnapshot
      ✔ should upload a snapshot
      ✔ should upload a snapshot without creds
      ✔ should upload a snapshot without flows
      ✔ should upload a snapshot without env (envVars: false)
      ✔ should upload a snapshot without env values (envVars: "keys")
```

### e2e tests added
image

### TODO:
- [x] Check / update docs where appropriate
- [x] Changelog: Ready: https://github.com/FlowFuse/flowfuse/pull/4610

## Related Issue(s)

#4605

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
@joepavitt added as reviewer for frontend.

This was rejigged slightly from the mock up screens in that the `credential secret` is now BELOW the option to enable disable import/export

The reason being: When clicking the component option "credentials", the new block of options were jumping up and down the dialog as the operation toggled the visibility of the  `credential secret`  field making it somewhat junky.

File: frontend/src/pages/application/Snapshots/components/dialogs/SnapshotExportDialog.vue [hunk 1]
```diff
@@ -1,29 +1,32 @@
...
-                        </div>
-                    </template>
-                </FormRow>
+            <form data-form="snapshot-export" @submit.prevent>
+                <ExportInstanceComponents v-model="parts" :error="errors.parts" header="Select the components to include in the snapshot" />
+                <template v-if="needsSecret">
+                    <FormRow containerClass="w-auto mt-6" :error="errors.secret" data-form="snapshot-secret">
+                        Secret
+                        <template #description>
+                            <p class="text-sm">A key used to encrypt any credentials in the snapshot.</p>
```
Comment (joepavitt) [inline frontend/src/pages/application/Snapshots/components/dialogs/SnapshotExportDialog.vue:10]:
```suggestion
                            <p class="text-sm">A key used to encrypt any credentials in the snapshot's flow.</p>
```

Fairly sure I'm right in saying that the `credentials` are in the `flow.json`? As an inexperienced NR user, I wouldn't necessarily know the difference here between env vars and credentials, I'd think credentials could be stored in Env Vrs. So want to add a little clarity here.

**Comment by Steve-Mcl** [comment]
> although front-end E2E's should be added,

Yeah, absolutely makes sense & agree. Will add front end tests.

**Comment by Steve-Mcl** [comment]
@joepavitt  e2e tests added

image

That was not easy - but definitely worth the effort as it turned up a couple of edge cases (like re-selecting same snapshot to upload would not work)

NOTE: There were a number of challenges in implementing the e2e tests for multiple downloads/uploads tests! One particular puzzling doozy was with uploading fixtures that several other users of cypress have faced over the years (and a hand full of "me too" issues on cypress issue tracker - none of which worked here!).  Essentially, `selectFile` call (for setting a fixture into a file input ready for upload) would not work in 2nd, 3rd, nth subsequent tests **unless** a different fixture file was used (that's why there are now `upload1.json`, `upload2.json`, ... files added in the fixtures dir - that took far to long to figure out and fix - bah!)

fingers crossed they pas on GH too 🤞

**Review by knolleary — APPROVED** [review]

**Comment by Steve-Mcl** [comment]
@joepavitt FYI

**Review by joepavitt — APPROVED** [review]
I'd already said it was all good in a comment, just hadn't clicked the Approve button/ With Nick's approval, you could have just gone ahead, but anyway, my explicit approval is given.
